A48-xds-least-request-lb-policy.md

```diff
@@ -0,0 +1,225 @@
+For `LEAST_REQUEST`, the policy name will be `"LEAST_REQUEST"`, and the config will be the one for the `least_request_experimental` LB Policy described below.
+When the `xds_cluster_resolver` policy sees `"LEAST_REQUEST"`, it will also assume that the
+locality-picking policy is `weighted_target` and the endpoint-picking policy is `least_request_experimental`.
+
+#### `least_request_experimental` LB Policy
+
+We will implement a `least_request_experimental` LB policy that uses the same
+algorithm as Envoy's implementation.  However, this policy will be
+implemented in a non-xDS-specific way, so that it can also be used without
+xDS in the future.
+
+##### LB Policy Config
+
+The `least_request_experimental` policy will have the following config:
+
+```proto
+message LeastRequestLoadBalancingConfig {
+  uint32 choice_count = 1;  // Optional, defaults to 2.
+}
+```
+
+The `choice_count` determines the number of randomly sampled subchannels which should be compared
+for each pick by the picker.
+Envoy currently supports any `choice_count` value greater than two.
+In the `least_request_experimental` policy however, only an effective `choice_count` in the
+range `[2, 10]` will be supported. If a `LeastRequestLoadBalancingConfig` with a `choice_count > 10`
+is received, the `least_request_experimental` policy will set `choice_count = 10`.
+
+##### Subchannel state handling
+
+The subchannel state handling implemented in the `round_robin` policy will be re-used for `least_request_experimental`.
+Implementations may break out this subchannel state handling and have `round_robin` and `least_request_experimental`
+inherit the logic instead. This would be similar to the "base balancer" already implemented in grpc-go
+[here](https://github.com/grpc/grpc-go/blob/03268c8ed29e801944a2265a82f240f7c0e1b1c3/balancer/base/balancer.go).
+The subchannel state handling used for the `round_robin` policy can be found in each language-specific implementation:
+
+* [Java](https://github.com/grpc/grpc-java/blob/c1e19af86dea0b9a8725969a95e116029397ad4d/core/src/main/java/io/grpc/util/RoundRobinLoadBalancer.java)
+* [C++](https://github.com/grpc/grpc/blob/79d684529d9db60aa2c5e82deb43e297a6818cfb/src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.cc)
+* [Go](https://github.com/grpc/grpc-go/blob/03268c8ed29e801944a2265a82f240f7c0e1b1c3/balancer/base/balancer.go)
+
+##### Aggregated Connectivity State
+
+The `least_request_experimental` policy will use the same heuristic for determining the aggregated
+connectivity state as the `round_robin` policy.
+The general rules in order of precedence are:
+
+1. If there is at least one subchannel with the state `READY`, the aggregated connectivity state is `READY`.
+2. If there is at least one subchannel with the state `CONNECTING` or `IDLE`,
+   the aggregated connectivity state is `CONNECTING`.
+3. If all subchannels have the state `TRANSIENT_FAILURE`, the aggregated connectivity state is `TRANSIENT_FAILURE`.
```

Comment:
So my interpretation of this is that the aggregated connectivity state can only transition `TRANSIENT_FAILURE -> READY` and not `TRANSIENT_FAILURE -> CONNECTING`.
How about adding something like this?

```suggestion
3. If all subchannels have the state `TRANSIENT_FAILURE`, the aggregated connectivity state is `TRANSIENT_FAILURE`.

Also note that since subchannels in this case only can transition from the state `TRANSIENT_FAILURE` to the state `READY` ,
the aggregated state also only may transition from the state `TRANSIENT_FAILURE` to the state `READY`.
```

Reply:
It looks like you've updated the text to refer to the "logical" subchannel state, as if that was an already well-understood term, but I don't think that most readers are going to know what that means.  I think it would be better to just say something like this:

For the purposes of the evaluating the aggregated connectivity state rules above, a subchannel that enters TRANSIENT_FAILURE should be considered to stay in TRANSIENT_FAILURE until it reports READY. If a subchannel is bouncing back and forth between TRANSIENT_FAILURE and CONNECTING while it attempts to reestablish a connection, we will consider it to be in state TRANSIENT_FAILURE for the purposes of the above aggregation rules.

Reply:
Thanks for the clarification!
Added a slightly modified version of your suggestion and removed the `"logic"` .